sdk/core/scheduler/main.cc

```diff
@@ -553,14 +599,45 @@ int futex_timed_wait(Timeout        *timeout,
 		return 0;
 	}
 	Thread *currentThread = Thread::current_get();
-	Debug::log("{} waiting on futex {} for {} ticks",
-	           currentThread,
+	Debug::log("Thread {} waiting on futex {} for {} ticks",
+	           currentThread->id_get(),
 	           address,
 	           timeout->remaining);
-	currentThread->futexWaitAddress = Capability{address}.address();
+	bool      isPriorityInheriting         = flags & FutexPriorityInheritance;
+	ptraddr_t key                          = Capability{address}.address();
+	currentThread->futexWaitAddress        = key;
+	currentThread->futexPriorityInheriting = isPriorityInheriting;
+	Thread  *blockingThread                = nullptr;
+	uint16_t blockingThreadID;
+	if (isPriorityInheriting)
+	{
+		// For PI futexes, the low 16 bits store the thread ID.
+		blockingThreadID = *address;
+		blockingThread   = get_thread(blockingThreadID);
+		// If we try to block ourself, that's a mistake.
+		if ((blockingThread == currentThread) || (blockingThread == nullptr))
+		{
+			return -EINVAL;
+		}
+		Debug::log("Thread {} boosting priority of {} for futex {}",
+		           currentThread->id_get(),
+		           blockingThread->id_get(),
+		           key);
+		blockingThread->priority_boost(priority_boost_for_thread(
+		  blockingThreadID, currentThread->priority_get()));
```

Comment:
Since `*address` is not marked `volatile` (and is marked `const`), it seems like there risk the compiler could replace `blockingThreadID` with `*address` in each occurrence.  I know we're running with IRQs off here but thinking about the impending DMA-capable future...

If we really need its ID, perhaps `blockingThread->id`, but I suspect we don't?

Reply:
We need its ID to work out which threads that are blocked on a futex are priority boosting that specific thread.

The compiler can issue two loads here only if it can prove that no intermediate stores are via a pointer that may alias this.  Since the pointer comes from outside and is not marked `restrict`, it cannot make that guarantee.